### ISSUE TITLE
fp16 parity issue with traced code on GPU

### ISSUE THREAD

**Issue by heglex** [pytorch/pytorch#109753]
### 🐛 Describe the bug

I have more details in [this colab notebook](https://colab.research.google.com/drive/1lf8RjX2p5F4E3MzN_blbRICughqHkSJP?usp=sharing), but the issue has to do with a specific sequence of operations that are not equivalent between the regular and traced models when running on a Tesla T4 GPU. The colab notebook was run with a T4 GPU runtime and `torch.__version__=="2.2.0.dev20230920+cu118"`.

Here is the code that fails. This does not fail on cpu, or if the operations in `CoordTransform` are simplified.

```python
import torch
print(torch.__version__)

gen = torch.random.manual_seed(123456789)
model_input = torch.rand(512, 512, 2, generator=gen, dtype=torch.float16).to(torch.device("cuda"))

class CoordTransform(torch.nn.Module):
    def forward(self, x_i: torch.Tensor) -> torch.Tensor:
        x = torch.arange(0, 512).to(x_i)
        x = x + x_i[0, :, 0]
        x = x * 0.4
        return x

model = CoordTransform()
ts_model = torch.jit.trace(model, model_input)

res = model(model_input)
res_ts = ts_model(model_input)

torch.testing.assert_close(res, res_ts, rtol=1e-4, atol=1e-1)
```

Output:
```
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
<ipython-input-20-931a9fd6043a> in <cell line: 6>()
      4 res = model(model_input)
      5 res_ts = ts_model(model_input)
----> 6 torch.testing.assert_close(res, res_ts, rtol=1e-4, atol=1e-1)

/usr/local/lib/python3.10/dist-packages/torch/testing/_comparison.py in assert_close(actual, expected, allow_subclasses, rtol, atol, equal_nan, check_device, check_dtype, check_layout, check_stride, msg)
   1518     if error_metas:
   1519         # TODO: compose all metas into one AssertionError
-> 1520         raise error_metas[0].to_error(msg)
   1521 
   1522 

AssertionError: Tensor-likes are not close!

Mismatched elements: 58 / 512 (11.3%)
Greatest absolute difference: 0.125 at index (282,) (up to 0.1 allowed)
Greatest relative difference: 0.00110626220703125 at index (282,) (up to 0.0001 allowed)
```

### Versions

Collecting environment information...
PyTorch version: 2.2.0.dev20230920+cu118
Is debug build: False
CUDA used to build PyTorch: 11.8
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.2 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: 14.0.0-1ubuntu1.1
CMake version: version 3.27.4
Libc version: glibc-2.35

Python version: 3.10.12 (main, Jun 11 2023, 05:26:28) [GCC 11.4.0] (64-bit runtime)
Python platform: Linux-5.15.120+-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 11.8.89
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: Tesla T4
Nvidia driver version: 525.105.17
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.9.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.9.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.9.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.9.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.9.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.9.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.9.0
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Address sizes:                   46 bits physical, 48 bits virtual
Byte Order:                      Little Endian
CPU(s):                          2
On-line CPU(s) list:             0,1
Vendor ID:                       GenuineIntel
Model name:                      Intel(R) Xeon(R) CPU @ 2.00GHz
CPU family:                      6
Model:                           85
Thread(s) per core:              2
Core(s) per socket:              1
Socket(s):                       1
Stepping:                        3
BogoMIPS:                        4000.33
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single ssbd ibrs ibpb stibp fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm mpx avx512f avx512dq rdseed adx smap clflushopt clwb avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves arat md_clear arch_capabilities
Hypervisor vendor:               KVM
Virtualization type:             full
L1d cache:                       32 KiB (1 instance)
L1i cache:                       32 KiB (1 instance)
L2 cache:                        1 MiB (1 instance)
L3 cache:                        38.5 MiB (1 instance)
NUMA node(s):                    1
NUMA node0 CPU(s):               0,1
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Mitigation; PTE Inversion
Vulnerability Mds:               Vulnerable; SMT Host state unknown
Vulnerability Meltdown:          Vulnerable
Vulnerability Mmio stale data:   Vulnerable
Vulnerability Retbleed:          Vulnerable
Vulnerability Spec store bypass: Vulnerable
Vulnerability Spectre v1:        Vulnerable: __user pointer sanitization and usercopy barriers only; no swapgs barriers
Vulnerability Spectre v2:        Vulnerable, IBPB: disabled, STIBP: disabled, PBRSB-eIBRS: Not affected
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Vulnerable

Versions of relevant libraries:
[pip3] numpy==1.23.5
[pip3] pytorch-triton==2.1.0+6e4932cda8
[pip3] torch==2.2.0.dev20230920+cu118
[pip3] torchaudio==2.2.0.dev20230920+cu118
[pip3] torchdata==0.6.1
[pip3] torchsummary==1.5.1
[pip3] torchtext==0.15.2
[pip3] torchvision==0.17.0.dev20230920+cu118
[pip3] triton==2.0.0
[conda] Could not collect

**Comment by ezyang** [comment]
any chance you can try torch.export on this instead

**Comment by ezyang** [comment]
torch.export.export

**Comment by heglex** [comment]
The following works:

`ts_model = torch.export.export(model, (model_input,))`

**Comment by heglex** [comment]
@ezyang One thing that doesn't look promising is the banner on the `torch.export` [docs](https://pytorch.org/docs/main/export.html) page that says:

> WARNING: This feature is a prototype under active development and there WILL BE BREAKING CHANGES in the future.

We also require the original `torch.jit.trace` for our codebase...

**Comment by ezyang** [comment]
No, I was mostly curious if the more sound torch.export was able to handle it correctly. It is cool that it is.

This is a legit bug with torch.jit.trace, but unless one of you wants to dig in and figure it out, we're unlikely to get a fix for it quickly. If you must use torch.jit.trace as the ingestion mechanism, one possibility would be to torch.export first, and pass the result along to torch.jit.trace (this should work, and remove the bug.) torch.export is prototype in 2.1 but it is the future.

**Comment by davidberard98** [comment]
Almost all of the jit.trace numerical issues I've seen have been due to the fusers' casting behavior. In eager mode, if you run `y = sin(x); z = cos(y)`, the intermediate value `y` will be stored in fp16. After fusing, `y` is kept in fp32. So, the result will likely be closer to the eager-fp32 result, but that means it can be different from the eager-fp16 result.

In a lot of the pytorch tests, we check this by checking (a) eager-fp16, (b) compiled-fp16 (either with jit or PT2), (c) eager-fp64. We check that `abs(a - c) < 2*abs(b - c)` (or something like this, I don't remember the specific factors).

If it's more important to you that you match eager-fp16 numerics than being closer to eager-fp32 numerics, you can turn off the fuser; but this can possibly reduce performance. Instructions on how to turn off the fuser are here:
https://github.com/pytorch/pytorch/blob/9567aaebe5385c2ce453c10c34180bba53773a93/torch/csrc/jit/OVERVIEW.md#fusers

**Comment by heglex** [comment]
Thank you @davidberard98, for the explanation and the workaround. Yes, it does seem to come down to whether the fuser tries to combine the add and multiply kernels. Given that there is some control over what happens here, we can use that to move forward.
